no-unused-modules hangs & overflows memory

In version 2.4.0

Running with the following rule:

    "import/no-unused-modules": [
        "error",
        {"unusedExports": true}
    ]

eslint hangs for a minute, and then node prints a dump:

> <--- Last few GCs --->
> 
> [13612:00000273A3998820]   138421 ms: Mark-sweep 1393.8 (1425.0) -> 1393.4 (1424.0) MB, 2458.2 / 0.0 ms  (average mu = 0.060, current mu = 0.008) allocation failure scavenge might not succeed
> [13612:00000273A3998820]   138429 ms: Scavenge 1394.6 (1424.0) -> 1394.2 (1426.5) MB, 5.7 / 0.0 ms  (average mu = 0.060, current mu = 0.008) allocation failure
> 
> 
> <--- JS stacktrace --->
> 
> ==== JS stack trace =========================================
> 
>     0: ExitFrame [pc: 00000367C635C5C1]
> Security context: 0x02cfc449e6e1 <JSObject>
>     1: visitSingle [000001E8CE77CE21] [C:\Users\Ivan Rubinson\Documents\Software Engineering\Work\PROJECTNAME\Front\node_modules\babel-eslint\node_modules\@babel\traverse\lib\context.js:~88] [pc=00000367C6C797DF](this=0x02e4eee45b41 <TraversalContext map = 000003F0895883F9>,node=0x02f0c5c7de69 <Node map = 000003F0895893C9>,key=0x02f78ec07d79 <String[2]: id>)
>   ...
> 
> FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
>  1: 00007FF79725ECE5
>  2: 00007FF797238196
>  3: 00007FF797238BA0
>  4: 00007FF7974C8D5E
>  5: 00007FF7974C8C8F
>  6: 00007FF797A069D4
>  7: 00007FF7979FD137
>  8: 00007FF7979FB6AC
>  9: 00007FF797A04627
> 10: 00007FF797A046A6
> 11: 00007FF7975A7767
> 12: 00007FF79763F44A
> 13: 00000367C635C5C1
> npm ERR! code ELIFECYCLE
> npm ERR! errno 134
> npm ERR! base-boilerplate@2.0.0 lint: `eslint ./client/src/`
> npm ERR! Exit status 134
> npm ERR!
> npm ERR! Failed at the base-boilerplate@2.0.0 lint script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     C:\Users\Ivan Rubinson\AppData\Roaming\npm-cache\_logs\2019-05-20T16_10_57_005Z-debug.log

The log from npm-cache is:

> 0 info it
> 
>  worked if it ends with ok
> 1 verbose cli [ 'C:\\Program Files\\nodejs\\node.exe',
> 1 verbose cli   'C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js',
> 1 verbose cli   'run',
> 1 verbose cli   'lint' ]
> 2 info using npm@6.4.1
> 3 info using node@v10.13.0
> 4 verbose run-script [ 'prelint', 'lint', 'postlint' ]
> 5 info lifecycle base-boilerplate@2.0.0~prelint: base-boilerplate@2.0.0
> 6 info lifecycle base-boilerplate@2.0.0~lint: base-boilerplate@2.0.0
> 7 verbose lifecycle base-boilerplate@2.0.0~lint: unsafe-perm in lifecycle true
> 9 verbose lifecycle base-boilerplate@2.0.0~lint: CWD: C:\Users\Ivan Rubinson\Documents\Software Engineering\Work\PROJECTNAME\Front
> 10 silly lifecycle base-boilerplate@2.0.0~lint: Args: [ '/d /s /c', 'eslint ./client/src/' ]
> 11 silly lifecycle base-boilerplate@2.0.0~lint: Returned: code: 134  signal: null
> 12 info lifecycle base-boilerplate@2.0.0~lint: Failed to exec lint script
> 13 verbose stack Error: base-boilerplate@2.0.0 lint: `eslint ./client/src/`
> 13 verbose stack Exit status 134
> 13 verbose stack     at EventEmitter.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\index.js:301:16)
> 13 verbose stack     at EventEmitter.emit (events.js:182:13)
> 13 verbose stack     at ChildProcess.<anonymous> (C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\lib\spawn.js:55:14)
> 13 verbose stack     at ChildProcess.emit (events.js:182:13)
> 13 verbose stack     at maybeClose (internal/child_process.js:962:16)
> 13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:251:5)
> 14 verbose pkgid base-boilerplate@2.0.0
> 15 verbose cwd C:\Users\Ivan Rubinson\Documents\Software Engineering\Work\PROJECTNAME\Front
> 16 verbose Windows_NT 10.0.17134
> 17 verbose argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "run" "lint"
> 18 verbose node v10.13.0
> 19 verbose npm  v6.4.1
> 20 error code ELIFECYCLE
> 21 error errno 134
> 22 error base-boilerplate@2.0.0 lint: `eslint ./client/src/`
> 22 error Exit status 134
> 23 error Failed at the base-boilerplate@2.0.0 lint script.
> 23 error This is probably not a problem with npm. There is likely additional logging output above.
> 24 verbose exit [ 134, true ]

--- a/src/rules/no-unused-modules.js
+++ b/src/rules/no-unused-modules.js
@@ -47,6 +47,7 @@ let preparationDone = false
 const importList = new Map()
 const exportList = new Map()
 const ignoredFiles = new Set()
+const filesOutsideSrc = new Set()
 
 const isNodeModule = path => {
   return /\/(node_modules)\//.test(path)
@@ -192,8 +193,9 @@ const getSrc = src => {
  * prepare the lists of existing imports and exports - should only be executed once at
  * the start of a new eslint run
  */
+let srcFiles
 const doPreparation = (src, ignoreExports, context) => {
-  const srcFiles = resolveFiles(getSrc(src), ignoreExports)
+  srcFiles = resolveFiles(getSrc(src), ignoreExports)
   prepareImportsAndExports(srcFiles, context)
   determineUsage()
   preparationDone = true
@@ -375,12 +377,17 @@ module.exports = {
         return
       }
 
-      // refresh list of source files
-      const srcFiles = resolveFiles(getSrc(src), ignoreExports)
+      if (filesOutsideSrc.has(file)) {
+        return
+      }
 
       // make sure file to be linted is included in source files
       if (!srcFiles.has(file)) {
-        return
+        srcFiles = resolveFiles(getSrc(src), ignoreExports)
+        if (!srcFiles.has(file)) {
+          filesOutsideSrc.add(file)
+          return
+        }
       }
 
       exports = exportList.get(file)

